### PR TITLE
MULE-10873: artifact plugin descriptors no longer unzips

### PR DESCRIPTION
**Pull request by fernandezlautaro**
@pablokraan: as we talked before, this is a subset of the PR done in https://github.com/mulesoft/mule/pull/3512/.

It keeps using the plugin.properties file, but it no longer unzips the plugins

### CODE REVIEW

File: modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorZipLoader.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Optional;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+/**
+ * Extracts the needed information of a mule plugin when its current format is a ZIP file.
```
Comment (pablokraan) [inline modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorZipLoader.java]: mule > Mule

File: modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorZipLoader.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.module.deployment.internal.plugin;
+
+import static java.lang.String.format;
+import static java.util.Optional.empty;
+import static org.apache.commons.lang3.StringUtils.removeEnd;
+import static org.mule.module.artifact.classloader.net.MuleArtifactUrlConnection.CLASSES_FOLDER;
```
Comment (pablokraan) [inline modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorZipLoader.java]: Why CLASSES_FOLDER is defined on MuleArtifactUrlConnection? (previous PR)

File: modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorZipLoader.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+ *
+ * @since 4.0
+ */
+class ArtifactPluginDescriptorZipLoader extends ArtifactPluginDescriptorLoader {
+
+  static final String EXTENSION_ZIP = ".zip";
+
+  private final ZipFile pluginZip;
+
+  ArtifactPluginDescriptorZipLoader(File pluginLocation) {
```
Comment (pablokraan) [inline modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorZipLoader.java]: Add javadoc

File: modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorZipLoader.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+  }
+
+  @Override
+  protected List<URL> getRuntimeLibs() throws MalformedURLException {
+    List<URL> urls = new ArrayList<>();
+    try {
+      Enumeration<? extends ZipEntry> entries = pluginZip.entries();
+      while (entries.hasMoreElements()) {
+        ZipEntry entry = entries.nextElement();
+        if (entry.getName().startsWith("lib/") && entry.getName().endsWith(".jar")) {
```
Comment (pablokraan) [inline modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorZipLoader.java]: Should reuse a constant for "lib"

Comment (pablokraan) [inline modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorZipLoader.java]: And another one for ".jar"

File: modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/plugin/DefaultArtifactPluginRepository.java [hunk 5]
```diff
@@ -55,65 +44,31 @@ public DefaultArtifactPluginRepository(ArtifactPluginDescriptorFactory pluginDes
...
    */
   private List<ArtifactPluginDescriptor> createApplicationPluginDescriptors() {
     List<ArtifactPluginDescriptor> pluginDescriptors = new LinkedList<>();
-
-    for (File pluginExpandedFolder : getContainerAppPluginsFolder()
-        .listFiles((FileFilter) DIRECTORY)) {
-      final ArtifactPluginDescriptor appPluginDescriptor = pluginDescriptorFactory.create(pluginExpandedFolder);
-      pluginDescriptors.add(appPluginDescriptor);
+    File[] pluginFiles = getContainerAppPluginsFolder()
+        .listFiles(file -> file.isDirectory() || endsWithIgnoreCase(file.getName(), ".zip"));
```
Comment (pablokraan) [inline modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/plugin/DefaultArtifactPluginRepository.java]: Can you defina a constant for ".zip" too?

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/application/ApplicationDescriptorFactoryTestCase.java [hunk 6]
```diff
@@ -48,11 +33,19 @@
...
 
-import org.junit.Before;
-import org.junit.Ignore;
-import org.junit.Rule;
-import org.junit.Test;
-import org.junit.rules.TemporaryFolder;
+import static java.util.Collections.emptyList;
+import static org.apache.commons.io.FileUtils.copyFile;
+import static org.apache.commons.io.IOUtils.copy;
+import static org.hamcrest.Matchers.*;
```
Comment (pablokraan) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/application/ApplicationDescriptorFactoryTestCase.java]: Fix imports

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/application/ApplicationDescriptorFactoryTestCase.java [hunk 7]
```diff
@@ -48,11 +33,19 @@
...
+import static org.junit.Assert.assertThat;
+import static org.junit.Assert.fail;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.mule.runtime.container.api.MuleFoldersUtil.*;
```
Comment (pablokraan) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/application/ApplicationDescriptorFactoryTestCase.java]: Fix imports

File: modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorLoader.java [hunk 8]
```diff
@@ -6,48 +6,129 @@
...
+import java.util.Properties;
+import java.util.Set;
 
 /**
  * Loads a {@link ArtifactPluginDescriptor} from a file resource.
  *
  * @since 4.0
  */
-public class ArtifactPluginDescriptorLoader {
+abstract class ArtifactPluginDescriptorLoader {
```
Comment (pablokraan) [inline modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorLoader.java]: There must be an abstract or a parameterized test case to test subclasses of this class. (similar to what ArtifactPluginDescriptorLoaderTestCase used to do)

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorLoaderTestCase.java [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+    new PluginDescriptorChecker(pluginFolder).exportingResources(parsedExportedResources)
+        .assertPluginDescriptor(pluginDescriptor);
+  }
+
+  @Test
+  public void parsesLibraries() throws Exception {
+    final File pluginFolder = createPluginFolder();
+
+    final File pluginLibFolder = new File(pluginFolder, "lib");
+    MatcherAssert.assertThat(pluginLibFolder.mkdir(), is(true));
```
Comment (pablokraan) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorLoaderTestCase.java]: Add static imports

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorLoaderTestCase.java [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+      return new ArtifactPluginDescriptorZipLoader(targetFile).load();
+    } else {
+      return new ArtifactPluginDescriptorFolderLoader(pluginLocation).load();
+    }
+  }
+
+  private static void compressDirectory(Path path, File targetFile) {
+    List<ZipResource> resources = new ArrayList<>();
+    try {
+      Files.walkFileTree(path, new SimpleFileVisitor<Path>() {
```
Comment (pablokraan) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorLoaderTestCase.java]: Use static imports

File: modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorLoaderTestCase.java [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+      File targetFile = new File(pluginLocation.getAbsolutePath() + EXTENSION_ZIP);
+
+      compressDirectory(pluginLocation.toPath(), targetFile);
+      return new ArtifactPluginDescriptorZipLoader(targetFile).load();
+    } else {
+      return new ArtifactPluginDescriptorFolderLoader(pluginLocation).load();
+    }
+  }
+
+  private static void compressDirectory(Path path, File targetFile) {
```
Comment (pablokraan) [inline modules/deployment/src/test/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorLoaderTestCase.java]: Move this function to ZipUtils class

File: modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorZipLoader.java [hunk 12]
```diff
@@ -27,16 +27,27 @@
 import java.util.zip.ZipFile;
 
 /**
- * Extracts the needed information of a mule plugin when its current format is a ZIP file.
+ * Extracts the needed information of a Mule plugin when its current format is a ZIP file.
  *
  * @since 4.0
  */
-class ArtifactPluginDescriptorZipLoader extends ArtifactPluginDescriptorLoader {
+public class ArtifactPluginDescriptorZipLoader extends AbstractArtifactPluginDescriptorLoader {
```
Comment (pablokraan) [inline modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorZipLoader.java]: ArtifactPluginDescriptorZipLoader > ZipArtifactPluginDescriptorLoader

File: modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorFolderLoader.java [hunk 13]
```diff
@@ -28,8 +30,14 @@
  *
  * @since 4.0
  */
-class ArtifactPluginDescriptorFolderLoader extends ArtifactPluginDescriptorLoader {
+public class ArtifactPluginDescriptorFolderLoader extends AbstractArtifactPluginDescriptorLoader {
```
Comment (pablokraan) [inline modules/deployment/src/main/java/org/mule/runtime/module/deployment/internal/plugin/ArtifactPluginDescriptorFolderLoader.java]: ArtifactPluginDescriptorFolderLoader > FolderArtifactPluginDescriptorLoader

File: tests/unit/src/main/java/org/mule/tck/ZipUtils.java [hunk 14]
```diff
@@ -69,4 +78,28 @@ public static void compress(File targetFile, ZipResource[] resources) {
...
+   * Takes a source {@link Path} it goes over every file recursively generating a temporary array of {@link ZipResource}s
+   * so that it can later compress it to the desired target.
+   *
+   * @param targetFile destination folder to compress the found files from {@code path}
+   * @param path to walk, while assembling a list of files to be later compressed.
+   */
+  public static void compressDirectory(File targetFile, Path path) {
+    List<ZipResource> resources = new ArrayList<>();
+    try {
+      Files.walkFileTree(path, new SimpleFileVisitor<Path>() {
```
Comment (pablokraan) [inline tests/unit/src/main/java/org/mule/tck/ZipUtils.java]: static imports!

Comment (fernandezlautaro) [inline tests/unit/src/main/java/org/mule/tck/ZipUtils.java]: this one was bc of the IDE, copying and pasting is bad.
Fixing it right now

**Pull request merged**